### DISCUSSION TITLE
Which GHC features are experimental?

### DISCUSSION THREAD

**Discussion by adamgundry** [ghc-proposals/ghc-proposals#635] [General]
Various recent proposals (e.g. #601, #617, #625) have made it clear that we should more clearly demarcate experimental from non-experimental features of GHC. Roughly speaking, non-experimental features are expected to remain stable between releases (with changes given careful consideration and ample deprecation warnings). Experimental features may change more frequently (without warning), and potentially users may be actively warned away from using them (#601) or need to jump through hoops to do so (#625).

Now this is inevitably a somewhat blurry distinction, not least because it depends on how we define "feature" and "experimental", and on the consequences of that decision (e.g. the strength of the expected stability guarantees for non-experimental features. But I think we should make an active effort to identify which features we collectively consider experimental. Eventually this topic probably needs a proposal in its own right, but for now, let's see if a rough consensus emerges. Please opine away...

**Comment by simonpj** [comment]
I think we need a draft to haggle over.   Can we just have a google sheet with a draft classification to get us going?  I'm sure a draft exists somewhere.

**Reply by adamgundry** [reply]
I previously started [a spreadsheet](https://docs.google.com/spreadsheets/d/1sRIcNKdflX2ogrx2EF3vKOn0vmv-UEOdladBqtYXdaw/edit#gid=0) with an exhaustive list of extensions, based on an earlier notion of classification. But I think we should frame this as about "features" in a broader sense than extensions. I've started to draft a [new document](https://docs.google.com/document/d/1B8E5XFrAmCcyH18OCBQBBOG7XIOw_SmPweKtpR7mJgY/edit) taking the latter approach. Please feel free to comment/edit/haggle.

**Comment by simonpj** [comment]
> I previously started [a spreadsheet](https://docs.google.com/spreadsheets/d/1sRIcNKdflX2ogrx2EF3vKOn0vmv-UEOdladBqtYXdaw/edit#gid=0) with an exhaustive list of extensions, based on an earlier notion of classification. But I think we should frame this as about "features" in a broader sense than extensions. I've started to draft a [new document](https://docs.google.com/document/d/1B8E5XFrAmCcyH18OCBQBBOG7XIOw_SmPweKtpR7mJgY/edit) taking the latter approach. Please feel free to comment/edit/haggle.

I'm a bit confused now.   

I think we want to know **for each extension**, whether it is regarded as Experimental.  A "stable package" (in the sense of #625) does not use those extensions.
* Are you saying that the spreadsheet is the current draft of that proposal?   (Or is it somehow supposed to be out-dated by the Google doc?) 
* The spreadsheet has many columns, where I was expecting one.  Maybe we should focus on column C Stability?  I'm not sure how to interpret the others.

I think that the google doc is intended to extend the "experimental" category to cover "features", which are not directly controlled by language extensions.   Good!   But
* It would be good to lay  out the purpose more clearly -- the current preamble is very global and appearse to cover language extensions too.  
* Linear Haskell, for example *is* controlled by a language extension.  Why is it needed here?  Ditto OverloadedRecordUpdate and RebindableSyntax.

**Reply by adamgundry** [reply]
The spreadsheet was something I started drafting in connection with (some hypothetical version of) #601, so that's why it has various extra columns/classifications. We can certainly make a more targeted copy if you like. But my suggestion is to use the document instead (covering both extensions and other features), extending it to give an argument for why each feature is to be considered experimental.

**Reply by simonpj** [reply]
I'm very keen to know
* for each extension (an exhaustive list thereof)
* is it experimental

The spreadsheet is good for this.  The google doc is not.  There are many extensions that are not covered by the google doc.  Could we have the spreadsheet for extensions, plus the google doc for broader questions, please?

**Reply by adamgundry** [reply]
Okay, that's fine. I've updated the spreadsheet to remove the old extra columns, and made a few revisions. Currently column C is a three-way classification (including a Deprecated category as well as Experimental and Stable); we could get rid of that but it seems like it's worthwhile keeping separate?

This is all still just my opinions, of course, so feel free to propose more changes. :smile:

**Reply by angerman** [reply]
As I'm just boarding a plane back to Singapore and had the chance to think about this whole topic a bit over the last week, I came to the realization that "experimental" might mean very different things to me.

Do we have an agreed upon definition of what "experimental" is?

My definition would be:
- experimental is for what the authors reserve the right to make changes without consideration for existing use.
- anything else is stable.

However this seems to be a _very_ different classification that others seem to consider, which seems to include theoretical and conceptual considerations, as opposed to the purely technical/mechanical classification I consider.

This also appears to be why my -experimental flag and "everything we currently have has to be assumed stable" are misunderstood? The exceptions might be LinearTypes and DependentHaskell as well as the few extensions that are _explicitly_ marked as experimental in the user guide.

Hence, if we can agree on a clear definition what experimental means (to everyone), classification might be easier?

**Reply by simonpj** [reply]
> experimental is for what the authors reserve the right to make changes without consideration for existing use.

That is close to what I would say, yes.   I say "close" because even for experimental features I would expect authors to give consideration to stability, but with a significantly lower bar than for unstable ones.

Essentially "stable" means "subject to #625" (which we have not quite merged yet for some reason.

Do others agree?

**Reply by hasufell** [reply]
I'm not sure I agree, because I expect there to be author bias.

So the classification should go through a channel that explicitly classifies, not merely lets authors of extensions make that decision. 

Likewise, even if an author thinks "well, I'm done"... that might not be how others (GHC SC?) sees it.

Or: the extension is done, but depends on theoretical foundations that are in flux.

In general, I don't find "author is likely going to make changes" much of an interesting classification. It should be possible to make changes anyway (e.g. through language editions or newly named variants).

I guess there are different motivations possible here. I think @angerman is purely thinking about upgrade cost due to heavily evolving extensions.

I'm more interested in semantics and soundness, of which I, as a non-academic, don't understand much! So this is where I see the main value.

But I concede, that could be expressed as yet another flag. However, you could also argue that under the motivation expressed by @angerman the flag should rather be `--no-unstable`, because it is only interested in change cadence and upgrade cost. Doesn't "experimental" mean more than that, especially in the context of language design?

**Reply by simonpj** [reply]
Maybe I should not have said "author".  I should have said "GHC Steering Committee", which is in charge of the classification.  Indeed that's what this thread is about.  

Uncertainty about semantics/soundness is a good reason for an extension to be experimental, because that means it might change.

I don't see this as difficult in principle:
* Stable = higher bar to change (see #625)
* Experimental = lower bar to change

That's about it.

**Reply by hasufell** [reply]
> Uncertainty about semantics/soundness is a good reason for an extension to be experimental, because that means it might change.

Well, but now the question is: can we classify a widely used extension, that hasn't been changed in a decade, as experimental, because it is indeed unsound?

From what I understand @angerman would say "no". Is that right?

**Reply by simonpj** [reply]
> Well, but now the question is: can we classify a widely used extension, that hasn't been changed in a decade, as experimental, because it is indeed unsound?

The committee will have to make judgement calls about hard cases.  By definition, hard cases are not black and white; there will be differing views.  But I doubt they will be resolved by trying to write increasingly precise language.  We will just have do debate each one on its merits.

Meanwhile, my suggestion above is that we don't allow the hard cases to de-rail the process.  Let's agree 90-95%, and then turn our attention to the hard cases.

**Reply by int-index** [reply]
The way I see it, this is all very easy and there's no need to overthink it.

* Stable extension: requires a proper migration process before any breaking change takes place (at least a warning for a few releases)
* Experimental extension: can change without a warning

That's it. This labeling is about our contract with the user. If someone uses an experimental extension, they should follow its development closely and prepare to adapt to changes. Users of stable extensions can rely on warnings to help them adapt to changes.

> and then turn our attention to the hard cases

What are the hard cases? I don't think there are any. We already have `LinearTypes` and `RequiredTypeArguments` marked as experimental in the User's Guide. Let's just mark everything else as stable. This is not a major commitment, it just means that breaking changes will require a warning.

**Reply by simonpj** [reply]
> Let's just mark everything else as stable

That would commit us to a very high stability bar on everything else, which I   For example I definitely regard ImpredicativeTypes as experimental.    ApplicativeDo and Arrows both badly need love; I'm not ready to say that they are "done".   OverloadedRecordUpdate is still in flux.

There is a spreadsheet with a draft classification linked up-thread.   Maybe take a look at that?

**Reply by int-index** [reply]
> That would commit us to a very high stability bar on everything else

I don't think so. My point is that the bar is not high. It's just that a breaking change must be preceded by a warning for a few releases. That's not that high of a bar, is it?

> There is a spreadsheet with a draft classification linked up-thread. Maybe take a look at that?

Right, I've missed it. Sure, maybe it's fine to mark a few more extensions as experimental if there aren't many users and there's ongoing work being done to improve them (perhaps slow work, but at least some changes every couple releases). Otherwise, I don't think it's fair to mark them "experimental" because there is no "experiment" taking place, only perhaps failed experiments that ended.

**Reply by tomjaguarpaw** [reply]
> a breaking change must be preceded by a warning for a few releases

How long do you imagine "a few releases" to be, in calendar time? To me it sounds like 18 or more months.

**Reply by simonpj** [reply]
>  It's just that a breaking change must be preceded by a warning for a few releases. That's not that high of a bar, is it?

It's higher than that.  As #625 says:

* GR1: don't make breaking changes in stable extensions at all
* GR2: but we can if there is a *compelling* reason to do so
* GR3: and if we must, then use publicity, deprecations etc to mitigate impact.

It even explicitly says "Note that (GR3) is not a reason to say that (GR1) is unimportant. A deprecation cycle defers costs; it does not reduce or eliminate them"   

So for stable features, the default position is: no breaking changes.  It is not "change what you like but add a deprecation cycle".

You could perhaps argue that we should change #625 but that's what it says right now and there has been plenty of opportunity to debate it.

**Reply by int-index** [reply]
> To me it sounds like 18 or more months.

Yes. That's why the "experimental" category is so useful — it allows much faster iteration. Basically, it tells the user "someone is working on it, expect changes", and at the same time gives more freedom and flexibility to the GHC contributor making those changes.

But if we're talking about a sporadic change once a few years, then adding 18 months to inform the users is only logical.

**Reply by int-index** [reply]
> So for stable features, the default position is: no breaking changes. It is not "change what you like but add a deprecation cycle".

I suppose I just don't see how this isn't true of _all_ changes. There's always a motivation, something that compels the proposal author to write the proposal and then implement it. If GR2 says "we can if there is a compelling reason to do so", it doesn't seem to say much at all. After all, no one is making changes just for the heck of it, right? Perhaps #625 simply codifies common sense, which is a good thing to do.

In other words, I still think it's rather simple:
* Stable extension: compelling reason required for a change, plus warnings to aid migration
* Experimental extension: compelling reason required for a change

**Reply by simonpj** [reply]
It isn't a black and white distinction, I grant you.  But for stable extensions, we are going to try hard not to make breaking changes; that is GR1.  For experimental extensions we don't try nearly so hard.    It's more than "add warnings" (GR3)

Moreover, #625 says "Even changes to Experimental extensions should seek to follow (GR3), but with a significantly lower "bar".", so whenever it is readily possible we should help users of experimental extensions by giving them a warning.

It's hard to debate this in the abstract.   Better to debate particular cases. So do look at the spreadsheet.

**Reply by int-index** [reply]
> It's hard to debate this in the abstract. Better to debate particular cases. So do look at the spreadsheet.

OK. Case study 1: **ApplicativeDo**. The spreadsheet proposes to mark it experimental, linking to
1. a proposal for a similar extension, ApplicativeComprehensions https://github.com/ghc-proposals/ghc-proposals/pull/526
2. a ticket about a program incorrectly rejected https://gitlab.haskell.org/ghc/ghc/-/issues/17768

Neither is a breaking change:

1. The proposal introduces a new language extension, so it does not affect ApplicativeDo per se.
2. Fixing the linked ticket will enable GHC to accept more programs, not fewer, so it's not a breaking change and is irrelevant to stability goals.

Furthermore, there is no one currently actively working on ApplicativeDo and there are no *planned* breaking changes. So I simply don't see how it's experimental. There is no experiment, and there is no reason to unnecessarily discourage users from using this neat language extension.

Case study 2: **Arrows**. Justification offered by the spreadsheet: longstanding bugs. Yet we should be able to fix bugs even in stable extensions, so that's irrelevant. Importantly, are there bugs where programs are _incorrectly accepted_ or with _incorrect runtime results_? Only those would require breaking changes to fix. It's certainly OK to fix bugs when correct programs are incorrectly rejected, as accepting them will only make GHC more permissive.

Furthermore, no one is _experimenting_ with Arrows. The only effort I'm aware of is https://github.com/ghc-proposals/ghc-proposals/pull/303, and it hasn't seen many updates in the last few years. Should we discourage stability-conscious users from Arrows because of that initiative, in the hope that the author (or someone else) will pick it up and actually change Arrows? I'm not sure. Maybe. But then that's what the spreadsheet should be citing.

Case study 3: **ImpredicativeTypes**. Do we expect any breaking changes, or is QuickLook working predictably enough? Who is planning to make those changes? Why should stability-conscious users avoid this extension? Will their programs break? I don't know, there's no justification in the spreadsheet.

Case study 4: **OverloadedRecordUpdate**. This time I totally agree, it's clearly experimental. We know there are changes pending (https://github.com/ghc-proposals/ghc-proposals/pull/583)! And we're still unsure about the design, so maybe there will be more changes after that.

Case study 5: **UnboxedTuples**. Again, how are those experimental? What are we experimenting with? Can anyone even come up with a program that is going to break because of some planned changes to unboxed tuples? They're used rather widely in performance-critical code, shouldn't we be careful about breaking it? Strange to see it marked experimental.


In general, I'd expect to see an extension marked as experimental in three cases

* there's someone in particular (a person or a team) working on improving the extension and we want to minimize the costs of iteration for them
* there are glaring design issues that require breaking changes to fix, and we want to minimize the activation energy for a new contributor to tackle the problem
  * if an extension is new, we might want to classify it as experimental because of _potential_ design issues that we expect early adopters to report

**Reply by angerman** [reply]
Thank you @int-index for going all the way to illustrate this with case studies!

FWIW the view @int-index has elaborated here very closely matches my understanding of what should be experimental or not. I also do not see this as very complicated but a rather simple black and white classification.

- Do we intent to change this featurw? yes/no? 
- Have we published this feature and it is already used in public? Yes/no?
- Is this a new extension? Yes/No?

I still see a different interpretation of the meaning of "experimental" here and this causes me concern.

If we want to "rethink" existing features, my suggestion is still to:

1.) create a new experimental extension/feature
2.) stabilize said new feature
3.) deprecate existing feature for new feature.

No in-place mutation.

**Reply by angerman** [reply]
Just to add. As @hasufell points out, to him (and I believe others), experimental does not mean not-stable. As such there can be features are experimental, yet stable.

This could be the crux here. To me something experimental, can not be stable.

I'm perfectly happy to run with a different term than experimental, if that's what needs to happen for this to be resolved. If experimental is an umbrella term for unsound, iffy, stable-but-not-pretty, something-we-should-sometime-rethink, ...

And we can all agree that what @int-index described is "unstable", I'm happy to use unstable as a term going forward. We do have the issue then that we currently call these things experimental in the user guide as well.

**Reply by simonpj** [reply]
> Just to add. As @hasufell points out, to him (and I believe others), experimental does not mean not-stable. As such there can be features are experimental, yet stable.

An experimental feature may have been stable for a while, and with no immediate plans for change; and yet, because we are concerned about aspects of the design, we may not want to tie ourselves to retaining accidental aspects of the current design.  So a design might be stable for now, but still labelled experimental because we don't want to commit to every aspect of the current design.

Examples

* ImpredicativeTypes.  I am very pleased with Quick Look.  And it has a peer-reviewed paper attached to it, which is good prima-facie evidence that it's not a total hack.    But it was the culmination of a 15-yr quest, and I'd be reluctant to commit to it being the final answer.  For example, there are [24 open tickets on ImpredicativeTypes](https://gitlab.haskell.org/ghc/ghc/-/issues/?sort=updated_desc&state=opened&label_name%5B%5D=ImpredicativeTypes&first_page_size=20) and I would not want to promise that the fixes would all be non-breaking.  For example, do we allow `C (forall a. a->a)` as a constraint?  Can we treat `forall a -> blah` as a tau-type (#24030)?   Some of these questions may be resolved in a non-breaking way, others may not.

* ApplicativeDo.  Again there is a paper (good). But again there are [24 open tickets](https://gitlab.haskell.org/ghc/ghc/-/issues/?sort=updated_desc&state=opened&label_name%5B%5D=ApplicativeDo&first_page_size=20) and I think we really might want a less complicated, but more predictable, feature.  So no immediate plans for change, but I am not confident that we have the canonical design.

* Arrows have been around for even longer.   But [GHC proposal 303](https://github.com/ghc-proposals/ghc-proposals/pull/303) argues that our typing rules are really not quite right.

* OverloadedStrings.  Pretty stable but there is some argument about bad behaviour on unicode.

For UnboxedTuples I agree that there are no big issues.  We can make that stable.

@angerman says

> If we want to "rethink" existing features, my suggestion is still to:
>
> 1.) create a new experimental extension/feature
> 2.) stabilize said new feature
> 3.) deprecate existing feature for new feature.

I am sympathetic to this, but I worry that it could impose very heavy overheads on the change.  Imagine we changed the typing rules for arrows (as above).  It would be extremely burdensome to implement two copies of the type checking and desugaring rules for arrows, and maintain both for a year or more.  Likewise we don't want OverloadedStrings and OverloadedStringsWithVerySlightlyDifferentBehaviourOnUnicode.  That would be confusing as well as burdensome to the implementation.

My suggestion: 
* make things experimental if we aren't confident in their design
* be willing to entertain arguments from the user community that currently-experimental extension X is so useful to them that it should be reclassified as stable, even though that risks freezing a non-final design

It's not black and white.  Even if we classify (say) OverloadedStrings as stable, we can still (GR2) make a breaking change; it's just that we have to argue a lot harder and put a lot more work into mitigation.

**Reply by int-index** [reply]
> So a design might be stable for now, but still labelled experimental because we don't want to commit to every aspect of the current design

And do we want to commit to _every_ aspect of the design of e.g. TypeFamilies?

**Reply by simonpj** [reply]
TypeFamilies are a particularly interesting example.   Their design is not satisfying, and we might want to change it.   And yet their use is so widespread that breaking changes (however well founded) will be disruptive.  So we might want to designate them stable, not because they truly are, but because we want to impose on ourselves the higher stability standard of  a stable extension, despite the costs thereof.

It's a case by case discussion,  But yes, I think we should label TypeFamilies stable for this reason.

**Reply by angerman** [reply]
I clearly have #617 in the back of my head, and that won't make any sense (to me), if we start classifying OverloadedStrings, and similar, maybe-not-great-but-widely-used, extensions as experimental. Hence me trying to figure out if we need a different term to describe truly experimental, as in this-is-a-preview-feature-we-would-like-to-get-feeback-on-but-make-absolutely-no-guarantees-about-it's-stability-or-future-availiabillity-whatsoever.

Maybe #617 needs to become `-preview`, instead of `-experimental`, and we can keep the term experimental to include things like "this is theoretically experimental, even though it's kinda stable, and used in the ecosystem, and we commit to not breaking it easily". As such `-no-preview` would then exclude only the very limited set of extensions/features for which the authors reserve the right to experiment and iterate at fast pace (as @int-index layer out), or for which we know they are unsound (and not used in the ecosystem yet!).

For a second iteration of `OverloadedStrings`, my suggestion would simply be `OverloadedStrings2`, which signals a second iteration of an existing feature, it also quite likely encourages people to migrate OverloadedStrings -> OverloadedStrings2 (because bigger numbers are always better 👍).

**Reply by simonpj** [reply]
> that won't make any sense (to me), if we start classifying OverloadedStrings, and similar, maybe-not-great-but-widely-used, extensions as experimental

Yes, as I say above, I think we may need to accept that some not-yet-settled features are so widely used that we should accept a higher stability bar for them, warts and all.

But I think think thing to do is simply to make that case, on a case-by-case basis!  I don't think that adding more categories will help us.  The important thing is that our categories are *actionable*.  In particular, Stable extensions have a higher bar for changes, and stronger expectations for deprecation cycles, than Experimental extensions.  That's the point of #625.   If we want another category we should say what different actions it implies.

Myself, I think Stable vs Experimental is plenty enough.  Ultimately the criterion is not "is it truly finished" (desirable though that might be), but rather "do we want to submit this extension to the high bar described by #625".

**Reply by Mikolaj** [reply]
I think the idea of `-preview` is to replace `-experimental` as a name, and similarly with the extension classification. I like that name better, because it more clearly speaks "do not use in a serious team project" and can't be as easily misinterpreted as "use freely if you are a competent/advanced/modern enough". As such it's more "actionable", easier to incorporate in a team policy and explain.

**Reply by angerman** [reply]
@Mikolaj No I'm proposing to have `-preview` in #617 instead of `-experimental`. The idea is that the compiler does not even know about LinearTypes, DependentHaskell and other actively developed features that are in the compiler for the "I'm feeling lucky" folks, but should be used in a serious team project, and the compiler should also not even consider _suggesting_ that extension. The behaviour is closer to: without `-preview` the compiler acts as if those features do not exists at all.

What we then have is that experimental is a superset of preview. It's all those extensions for which we consider them to on a broader "experimental" level. I for one can not see how OverloadedStrings, or anything remotely used in public code on hackage could end up in the set of `preview` features. At the same time I want to give those people that want to experiment with bleeding edge features the ability to opt in to `-preview` for their toy projects, to try out the latest linear types, or what not. But this is clearly with the expectation that what ever code they write, will likely need to be adjusted, and they need to closely follow the preview development.

I initially wanted to use `-experimental` for that, but as it appears, the understood notion of experimental is much looser than the strict one I intended.

If we have ApplicativeDo, Arrows, ImpredicativeTypes, OverloadedRecordUpdate, UnboxedTuples, OverloadedStrings, ... in the experimental set, a compiler that by default disables all of these, has no practical value today. You will not be able to compile any non-trivial haskell program, as you'll likely hit at least one of these in your dependency tree somewhere.

**Reply by angerman** [reply]
And no, I do not consider features being behind a language extension to be enough, because they are still available _by default_. And the compiler will _know_ about them, and treat them all equally. That includes it will helpfully suggest to you those features.

I guess what we'll end up with is:
```
.--------.--------------------------. 
|        |              .---------. |
| Stable | Experimental | Preview | |
|        |              '---------' |
'--------'--------------------------'
```
Question: is there an overlap between Stable and Experimental? Can an extension be Stable _and_ Experimental? @hasufell seems to think so, while I believe @simonpj sees them as disjunct?


EDIT: Note, as of today, the compiler knows nothing about any of those. It consider _all_ of them the _same_. Stable, and recommendable to anyone using the compiler.

EDIT2: @hasufell correctly pointed out, that I made a mistake, and Stable should not cover experimental, at least not fully.

**Reply by Mikolaj** [reply]
Oh, I see. In that case, my point is, I don't care whether the Experimental layer in your picture is implemented, but I do care that the Preview is, because it has a better name and a better semantics (more actionable, since you can live without the features, so you can actually avoid them in practice, just as you explain). Are we able to agree and implement both at the same time? If not, why not start with Preview only? Frankly, what's the point of Experimental once we have Preview? Just the lack of stability guarantees? Are there better ways to express those?

In any case, I didn't want to restart the discussion, just voice my opinion that what teams might really care about is a narrow enough class with a suggestive name (automation is welcome, but more for individual new users than teams, I'd imagine). Fine-grained layers of non-actionable distinctions may be as readily ignored as any documentation ever is, even with automation (e.g., just set the `--experimental` flag always, on autopilot or in a global wrapper script, as copy-pasted from the net without understanding).

**Reply by angerman** [reply]
@Mikolaj #617 does come with an implementation. IOG had someone implement at least the basics for it, and the transitive tainting (dependencies compiled with `-preview` would be tainted), so that you know for certain that your dependencies do not include any preview features.

Thus, if we can get agreement on #617 (after I change the naming), it's probably a fairly quick way to have it implemented. However do note that #617 says moving extensions into preview will be a second proposal. So we'd start out with an initially empty set. However adding items into preview is doable, and if the general course is agreed upon, I don't think hard.

**Reply by simonpj** [reply]
I'm afraid that I have no idea what the difference between Experiemental and Preview is.  If you are proposing to pause this thread #635, and first want to propose a new category, could you write a spec for the categories, and how they affect #625?   Thanks!

To be clear, I am proposing just two buckets (I don't care what they are called):
* Stable: high bar for breaking changes
* Experimental: lower bar for breaking changes

**Reply by Mikolaj** [reply]
I think from a commercial user perspective, a more useful metric than "likelihood of breaking changes" would be that divided by "usefulness" where, if a feature is mandatory (everybody uses is directly or transitively), "usefulness" is infinity. Basically, cost divided by benefit. That's how type families don't get flagged as unsuitable for commercial use despite being unstable.

**Reply by adamgundry** [reply]
I find @angerman's picture in https://github.com/ghc-proposals/ghc-proposals/discussions/635#discussioncomment-8380131 very helpful. Perhaps we should indeed have three buckets:
 * Stable: clearly-specified, high bar for breaking changes, recommended to anyone using the compiler (e.g. `GADTs`)
 * Experimental: may be under-specified, lower bar for breaking changes, but still available to users without massive ceremony (e.g. `ImpredicativeTypes`)
 * Preview: need not have a specification at all, breaking changes without warning, users actively discouraged from using perhaps via a special flag as in #617 (e.g. `OverloadedRecordUpdate`)

The boundary between Preview and everything else can be sharp and machine-checked. But the boundary between Stable and Experimental is inherently somewhat fuzzy (e.g. combining `FunctionalDependencies` with overlapping instances puts us into Experimental territory, but we want to keep the core behaviour of both features relatively stable).

EDIT: another way to look at this is comparable to the distinction between `base` (stable), `ghc-experimental` (experimental but user-facing) and `ghc-internal` (here be dragons).

**Reply by jappeace** [reply]
Please, just keep it simple for now, we can add a preview bucket later. :sweat_smile:

**Reply by maralorn** [reply]
My impression is that in this picture (where --preview is more or less that what I previously assumed to be --experimental, aka the don't use this unless you and the users of your code really are okay with breakage)
the new and wider "experimental" category seems to largely coincide with "everything we don't want in a language edition." (With maybe the exception of the few stable-but-not-reasonably-on-by-default extensions.)

**Comment by hasufell** [comment]
I'm very interested in the classification of "experimental" vs "non-experimental".

One private discussion that sparked confusion and disagreement with @angerman was: under what principles is this distinction made *exactly*. How these extensions are classified also affects https://github.com/ghc-proposals/ghc-proposals/pull/617

Let's say for example: what if we consider `OverloadedStrings` as experimental (because it has [questionable semantics on ByteStrings](https://github.com/haskell/bytestring/issues/140))? That would mean that most of hackage today uses "experimental features".

**Do we consider "usage" and "ecosystem effects" when deciding this?** I think, in some way, the GHCXXXX language editions do that too (I followed the practice of the last language editions discussions a bit). But I personally would like to see very "aggressive" classification of (non-)experimental, because:

- it opens the door to fixing these extensions quickly (because as per https://github.com/ghc-proposals/ghc-proposals/pull/625 ...stable packages don't use experimental features per definition... so it's ok to "break" this extension)
- starting with a large set of experimental extensions seems like a good way to put positive pressure on extension development and stabilize them

In that sense, yes: I want `OverloadedStrings` to be classified as experimental (and then make the next GHC version reject bytestring literals).

Interesting is also: an extension is not per se experimental... it may now depend on the GHC version. How can packages say "I want to depend only on the stable version of this extension"? I think https://github.com/ghc-proposals/ghc-proposals/pull/617 may come into effect here again.

**Reply by adamgundry** [reply]
I don't think we have a clear consensus on the principles for classifying experimental vs. non-experimental. My hope is that actually trying to produce a classification will help make that clearer.

I think we need the classification to reflect the actual likelihood of features requiring substantial breaking changes. (Though of course it is hard to predict the future.) Aggressively defining things as experimental which are de facto stable (e.g. `OverloadedStrings`) seems problematic to me: it will cause a lot of disruption if we break them, and sticking an "experimental" label on them won't help reduce the impact.

**Reply by hasufell** [reply]
Right, so we may have a third class, which is "stable, but unsound".

And it seems you're arguing for not breaking/fixing it, but do what instead? Create a new better extension and then introduce a long deprecation cycle?

**Reply by adamgundry** [reply]
Yes, I think we want a third class "deprecated" meaning "stable, but discouraged". The spreadsheet includes Deprecated as a separate category, and I've updated the document to discuss some of this. GHC can (indeed, often already does) issue warnings when a deprecated extension/feature is used.

**Reply by simonpj** [reply]
> Yes, I think we want a third class "deprecated" meaning "stable, but discouraged".

I'm fine with having that.  

But it doesn't address the transitional problem that Julian identifies: what about extensions that are widely used but still have problems, like OverloadedStrings. 

My instinct is this:

* Let's first do all the easy classification.   That should deal with 90% of the cases, and that is progress.
* Then we can identify the tough cases and discuss them.

But let's not de-rail the entire process because of one of two tricky cases.

**Reply by konsumlamm** [reply]
IMO `ByteString` having a problematic `IsString` instance isn't a problem of `OverloadedStrings`, but of `bytestring`. That's like saying `Monad` is bad, because some type has a bad `Monad` instance.

**Reply by hasufell** [reply]
@konsumlamm I disagree. The fact that `OverloadedStrings` uses `IsString` instances is an implementation detail.

These are very different relationships and your analogy doesn't make sense to me.

It could be implemented without a class, it could throw compile time warnings and a host of other things.

You can call it a design issue if you like. But also note that the documentation of the extension **explicitly mentions ByteString support**.

**Comment by simonpj** [comment]
(Side note: this thread is a "discussion" which is different to an "issue" or a "pull request".  I didn't know that "discussions" even existed!  For what it's worth, I think this particular thread should be an "issue".)

**Comment by adamgundry** [comment]
Note that @telser has followed up on this discussion with a proposal to classify certain extensions, on which more input would be welcome: https://github.com/ghc-proposals/ghc-proposals/pull/669
